pytorch_lightning/plugins/collective/torch_collective.py

```diff
@@ -0,0 +1,113 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Optional, Union
+
+import torch
+import torch.distributed
+
+from pytorch_lightning.overrides.torch_distributed import broadcast_object_list
+from pytorch_lightning.plugins.collective import Collective
+from pytorch_lightning.utilities import _TORCH_GREATER_EQUAL_1_8
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.distributed import all_gather_ddp_if_available, distributed_available
+from pytorch_lightning.utilities.distributed import group as _group
+from pytorch_lightning.utilities.distributed import ReduceOp, sync_ddp_if_available
+from pytorch_lightning.utilities.types import _METRIC_COLLECTION
+
+
+class TorchCollective(Collective):
+    """Collective interface for DDP, DDPSpawn, DP and DDP2."""
+
+    def __init__(
+        self,
+        local_reduce: bool = False,
+        rank: Optional[int] = None,
+        device: Optional[Union[str, torch.device]] = torch.device("cpu"),
+        device_id: Optional[int] = None,
+        world_size: int = 1,
```

Comment:
Since `Collective` doesn't own a pg (a pg can be passed in), it shouldn't maintain a `rank` and a `world_size`. A process can be a member of process groups of different size, and be different ranks in different groups.

Reply:
If `Collective` is meant to represent a communicator instance whose membership includes all processes that runs Lightning trainer, it should own or be assigned to a fixed reference to a process group.

Reply:
@yifuwang since it previously came up around whether the trainer should teardown the global process group - do you think the collective should be responsible for the setup/teardown of this? or do you think that logic should remain in the training type plugin, while the `Collective` can be called with any pg/rank/world size passed in?

Reply:
> since it previously came up around whether the trainer should teardown the global process group - do you think the collective should be responsible for the setup/teardown of this?


It'll still depend on whether or `Lightning` set up the global process group. It's not something `Lightning` can control. More and more model parallel use cases need to set up the global pg before even initializing the trainer (e.g. `torchrec`). The use of ` ShardedTensor` depends on a pg as well, which likely means that it'll be the case too for the `torch.distributed` version of `FSDP`.

So we can do what we are doing today: only shutdown the global pg if we created it.

> or do you think that logic should remain in the training type plugin, while the Collective can be called with any pg/rank/world size passed in?


If somehow a caller already has access to the raw process group, I don't think it makes any sense to go through `Collective` anymore, whose sole purpose is to hide the difference between different communication backends. In that case, they should just use `torch.distributed` APIs directly.

Reply:
Does lightning support multiple process group now?
How about we address in a different issue `support multiple process group in lightning`